### PR TITLE
Enable pbf output for isochrone endpoint

### PR DESCRIPTION
**Pull request by chrstnbwnkl**
# Issue

Adds pbf output option for the isochrone endpoint. 

The way the geometries are represented in the proto message right now is pretty simple, but analogous to how the outputs are serialized to GeoJSON: each contour is treated as a ring and its left to the end user to figure out  the exact relationships between them. In most cases, this is easy because there is only one exterior ring and possibly holes. But I can identify at least two scenarios where the response contains a feature (or `contour` message in pbf) with more than one exterior ring:

  1. multi-locations, if their expansions do not intersect
  2. when the expansion includes a ferry edge
  3. possibly when using transit (haven't tested this yet but assume it's the same as with ferry, because you can only enter or leave at a specific stop, not along the way)

In these case, multiple exterior rings are produced, as well as inner rings that would have to be associated to the right outer one.

In any case, maybe there is a better way of representing the output geometries that includes taking some of the heavy lifting off the users by associating inner and outer rings correctly and construct a multipolygon feature for each contour. (EDIT: see proposal in #4575 )
  

## Tasklist

 - [x] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Update the docs with any new request parameters or changes to behavior described
 - [x] Update the [changelog](CHANGELOG.md)
 - [ ] If you made changes to the lua files, update the [taginfo](taginfo.json) too.

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: src/thor/isochrone_action.cc [hunk 1]
```diff
@@ -46,11 +46,11 @@ std::string thor_worker_t::isochrones(Api& request) {
   // we have parallel vectors of contour properties and the actual geojson features
   // this method sorts the contour specifications by metric (time or distance) and then by value
   // with the largest values coming first. eg (60min, 30min, 10min, 40km, 10km)
-  auto isolines =
-      grid->GenerateContours(contours, options.polygons(), options.denoise(), options.generalize());
+  auto contours =
```
Comment (chrstnbwnkl) [inline src/thor/isochrone_action.cc:49]: renaming for consistency

File: src/tyr/isochrone_serializer.cc [hunk 2]
```diff
@@ -12,16 +12,16 @@ using namespace valhalla::baldr::json;
...
+using namespace valhalla;
+using namespace tyr;
 
-std::string serializeIsochrones(const Api& request,
-                                std::vector<midgard::GriddedData<2>::contour_interval_t>& intervals,
-                                midgard::GriddedData<2>::contours_t& contours,
-                                bool polygons,
-                                bool show_locations) {
+std::string
+serializeIsochroneJson(Api& request,
```
Comment (chrstnbwnkl) [inline src/tyr/isochrone_serializer.cc:20]: separate json and pbf construction

File: src/tyr/isochrone_serializer.cc [hunk 3]
```diff
@@ -142,5 +142,68 @@ std::string serializeIsochrones(const Api& request,
...
+    con.set_metric_value(std::get<1>(interval));
+
+    // for each feature
+    for (const auto& feature : contour) {
+      // for each ring/contour
+      for (const auto& ring : feature) {
+        // construct a geometry
+        Isochrone_Geometry geom;
+        for (const auto& pair : ring) {
+          geom.add_coords(round(pair.lng() * 1e6));
```
Comment (chrstnbwnkl) [inline src/tyr/isochrone_serializer.cc:171]: coords are just scaled for now, not delta encoded

File: src/thor/isochrone_action.cc [hunk 4]
```diff
@@ -17,13 +17,13 @@ std::string thor_worker_t::isochrones(Api& request) {
   auto costing = parse_costing(request);
 
   // name of the metric (time/distance, value, color)
-  std::vector<GriddedData<2>::contour_interval_t> contours;
+  std::vector<GriddedData<2>::contour_interval_t> intervals;
```
Comment (nilsnolde) [inline src/thor/isochrone_action.cc:20]: yeah thanks for that!

File: src/tyr/isochrone_serializer.cc [hunk 5]
```diff
@@ -142,5 +142,66 @@ std::string serializeIsochrones(const Api& request,
...
+  Isochrone& isochrone = *request.mutable_isochrone();
+
+  // construct contours
+  for (size_t isoline_index = 0; isoline_index < contours.size(); ++isoline_index) {
+    const auto& contour = contours[isoline_index];
+    const auto& interval = intervals[isoline_index];
+
+    valhalla::Isochrone::Contour con;
+    con.set_metric(std::get<2>(interval) == "time" ? Isochrone_metric_type_time
+                                                   : Isochrone_metric_type_distance);
```
Comment (nilsnolde) [inline src/tyr/isochrone_serializer.cc:158]: nit: short form is somewhat nicer IMO
```suggestion
    con.set_metric(std::get<2>(interval) == "time" ? Isochrone::time
                                                   : Isochrone::distance);
```

File: src/tyr/isochrone_serializer.cc [hunk 6]
```diff
@@ -142,5 +142,66 @@ std::string serializeIsochrones(const Api& request,
...
+    for (const auto& feature : contour) {
+      // for each ring/contour
+      for (const auto& ring : feature) {
+        // construct a geometry
+        Isochrone_Geometry geom;
+        for (const auto& pair : ring) {
+          geom.add_coords(round(pair.lng() * 1e6));
+          geom.add_coords(round(pair.lat() * 1e6));
+        }
+        con.add_geometries()->CopyFrom(geom);
```
Comment (nilsnolde) [inline src/tyr/isochrone_serializer.cc:172]: not sure if `CopyFrom` is best here.. sounds like it's copying;) there's also a `SwapSmth` function(s) or change the order of things by creating a contour at the top with `auto* con = isochrone.mutable_contours->Add()`  and then a `auto* geom = con.mutable_geometries().Add()` where you can do `geom->add_coords()`.

File: src/valhalla_run_isochrone.cc [hunk 7]
```diff
@@ -177,7 +177,7 @@ int main(int argc, char* argv[]) {
 
   // Compute the isotile
   auto t1 = std::chrono::high_resolution_clock::now();
-  Isochrone isochrone;
+  valhalla::thor::Isochrone isochrone;
```
[the remark this answers is not in the document]
Comment (nilsnolde) [reply src/valhalla_run_isochrone.cc:180]: hm, yeah, it'd be much nicer if we had a `valhalla::pbf` namespace, this is always a problem

**Review by nilsnolde — CHANGES_REQUESTED** [review]
Nice, thanks for this! Just a few nits from my side and one minor problem in the pbf serializer.

File: src/tyr/isochrone_serializer.cc [hunk 8]
```diff
@@ -153,27 +153,23 @@ std::string serializeIsochronePbf(Api& request,
...
     for (const auto& feature : contour) {
       // for each ring/contour
       for (const auto& ring : feature) {
         // construct a geometry
-        Isochrone_Geometry geom;
+        auto* geom = con->mutable_geometries()->Add();
         for (const auto& pair : ring) {
-          geom.add_coords(round(pair.lng() * 1e6));
-          geom.add_coords(round(pair.lat() * 1e6));
+          geom->add_coords(round(pair.lng() * 1e6));
```
Comment (nilsnolde) [inline src/tyr/isochrone_serializer.cc:168]: Oh right, I didn’t see that initially. I guess it’s fine to mix lat/lon in the same array. Another concept would be to do lats and lons in separate arrays, but I don’t have a strong opinion on it. The latter is possibly a tiny bit more trivial to parse out for a client.

**Review by nilsnolde — DISMISSED** [review]

File: src/tyr/isochrone_serializer.cc (hunk 8 above)
Comment (chrstnbwnkl) [reply src/tyr/isochrone_serializer.cc:168]: I mean can it get any more trivial :smile: we can provide examples somewhere in the docs if necessary
```cpp
for (int i = 0; i < coords.size(); i += 2)
```

**Review by nilsnolde — DISMISSED** [review]
:ship:

File: src/tyr/isochrone_serializer.cc [hunk 9]
```diff
@@ -142,5 +142,62 @@ std::string serializeIsochrones(const Api& request,
...
+namespace tyr {
+
+std::string serializeIsochrones(Api& request,
+                                std::vector<midgard::GriddedData<2>::contour_interval_t>& intervals,
+                                midgard::GriddedData<2>::contours_t& contours,
+                                bool polygons,
+                                bool show_locations) {
+
+  // if the user requested pbf output
+  if (request.options().format() == Options_Format_pbf) {
```
Comment (chrstnbwnkl) [inline src/tyr/isochrone_serializer.cc:189]: not sure how I missed this one... if #4575 gets discarded in favor of this one, this needs to be removed

Comment (nilsnolde) [reply src/tyr/isochrone_serializer.cc:189]: oops, didn't see that either.. wouldn't break anything, but yeah it's redundant

Comment (nilsnolde) [reply src/tyr/isochrone_serializer.cc:189]: I'd say we should merge this first then merge into #4575 to reduce the changes

File: src/tyr/isochrone_serializer.cc [hunk 10]
```diff
@@ -142,5 +142,62 @@ std::string serializeIsochrones(const Api& request,
...
+    return serializeIsochronePbf(request, intervals, contours);
+  }
+
+  switch (request.options().format()) {
+    case Options_Format_pbf:
+      return serializeIsochronePbf(request, intervals, contours);
+    case Options_Format_json:
+      return serializeIsochroneJson(request, intervals, contours, polygons, show_locations);
+    default:
+      throw;
```
Comment (nilsnolde) [inline src/tyr/isochrone_serializer.cc:199]: what is the bare `throw` doing here actually? it seems off, there's nothing catching any exception before this

File: proto/isochrone.proto [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+message Isochrone {
+
+  enum metric_type {
+    time = 0;
+    distance = 1;
+  }
+
+  message Geometry {
+    repeated sint32 coords = 1;
```
Comment (chrstnbwnkl) [inline proto/isochrone.proto:14]: put packed = true

**Comment by chrstnbwnkl** [comment]
Closing in favor of #4575

**Pull request closed without merging**